static/js/publisher/market/market.js

```diff
@@ -12,6 +12,91 @@ function initSnapIconEdit(iconElId, iconInputId) {
   });
 }
 
+function initSnapScreenshotsEdit(screenshotsToolbarElId, screenshotsWrapperElId, initialState) {
+  // DOM elements
+  const screenshotsToolbarEl = document.getElementById(screenshotsToolbarElId);
+  const screenshotsWrapper = document.getElementById(screenshotsWrapperElId);
+
+  // simple state handling (and serializing as JSON in hidden input)
+  const state = {};
+  const stateInput = document.createElement('input');
+  stateInput.type = "hidden";
+  stateInput.name = "state";
+
+  screenshotsToolbarEl.parentNode.appendChild(stateInput);
+
+  const setState = function(nextState) {
+    for (let key in nextState) {
+      if (nextState.hasOwnProperty(key)) {
+        state[key] = nextState[key];
+      }
+    }
+
+    stateInput.value = JSON.stringify(state);
+  };
+
+  setState(initialState);
+
+  // templates
+  const screenshotTpl = (screenshot) => `
+    <div class="col-2">
+      <img src="${screenshot.url}" alt="" />
+    </div>
+  `;
+
+  const emptyTpl = () => `
+    <div class="col-12">
+      <a class="p-empty-add-screenshots js-add-screenshots">Add images</a>
+    </div>
+  `;
+
+  const renderScreenshots = (screenshots) => {
+    if (screenshots.length) {
+      screenshotsWrapper.innerHTML = screenshots.map(screenshotTpl).join("");
+    } else {
+      screenshotsWrapper.innerHTML = emptyTpl();
+    }
+  };
+
+  const render = () => {
+    renderScreenshots(state.images.filter(i => i.type === "screenshot"));
+  };
+
+  render();
+
+  const onScreenshotsChange = function() {
+    const fileList = this.files;
+
+    for (let i = 0; i < fileList.length; i++) {
+      const file = fileList[i];
+      setState({
+        images: state.images.concat([{ file, url: URL.createObjectURL(file), name: file.name, type: "screenshot" }])
+      });
+
+      render();
```

Comment:
Don't you want to render after the `for loop`?

Reply:
Right, good catch.